File: python/src/caspian/interpreters/process.py
"""Process interpreter — the ONE inbound pipeline.

verify → parse → step → HostPort → execute → transport.

Webhook, poll, and hosted gateway all call handle_webhook. This module must
not import the B-surface package (import-linter + skill layering).
"""

from __future__ import annotations

from typing import Protocol

from caspian.core.commands import Command, Host, Typing
from caspian.core.errors import DecodeError
from caspian.core.overlap import OverlapDecision, drain_transition
from caspian.core.ports import AdapterPort, Connection, HostPort, RawInbound, Result, Sent
from caspian.core.predicates import evaluate
from caspian.core.step import StepResult, StepState, step
from caspian.core.types import App, Event, Rule


class Transport(Protocol):
    def dispatch(self, sent: Sent) -> Result: ...


class ProcessInterpreter:
    """Runner for one connection. Overlap state and the latest queued event live here."""

    def __init__(
        self,
        app: App,
        adapter: AdapterPort,
        connection: Connection,
        *,
        host: HostPort | None = None,
        transport: Transport | None = None,
    ) -> None:
        self._app = app
        self._adapter = adapter
        self._connection = connection
        self._host = host
        self._transport = transport
        self._state = StepState()
        self._pending: dict[str, Event] = {}

    def handle_webhook(self, raw: RawInbound) -> list[Result]:
        """Process one raw inbound payload end-to-end. Returns per-command results."""
        verify = getattr(self._adapter, "verify", None)
        if callable(verify) and not verify(raw, self._connection):
            return [Result.err(DecodeError(reason="Webhook signature verification failed"))]

        parse_result = self._adapter.parse(raw)
        if not parse_result.is_ok:
            return [parse_result]

        events: list[Event] = parse_result.value
        results: list[Result] = []
        admitted: list[tuple[Event, StepResult, str]] = []

        for event in events:
            ack = getattr(self._adapter, "acknowledge", None)
            if callable(ack):
                ack_result = ack(event, self._connection)
                if ack_result is not None:
                    results.append(self._maybe_dispatch(ack_result))

            key = self._key(event)
            sr = step(
                self._state,
                event,
                self._app,
                channel_name=self._channel_of(event),
                overlap_key=key,
            )
            if sr.dropped:
                continue
            if not sr.commands:
                self._pending[key] = event
                continue
            admitted.append((event, sr, key))

        for event, sr, key in admitted:
            results.extend(self._execute(event, sr))
            results.extend(self._drain(key, sr.matched_rule))

        return results

    def _key(self, event: Event) -> str:
        return self._adapter.overlap_key(event)

    def _channel_of(self, event: Event) -> str:
        channel_of = getattr(self._adapter, "channel_of", None)
        if callable(channel_of):
            return str(channel_of(event))
        return self._adapter.name

    def _match(self, event: Event) -> Rule | None:
        channel_name = self._channel_of(event)
        for rule in self._app.rules:
            if evaluate(rule.predicate, event, channel_name=channel_name):
                return rule
        return None

    def _execute(self, event: Event, sr: StepResult) -> list[Result]:
        commands: list[Command] = []
        for cmd in sr.commands:
            if isinstance(cmd, Host):
                if self._host is not None:
                    commands.extend(
                        self._host.run(
                            cmd.handler_id, event, skipped_count=sr.skipped_count
                        )
                    )
            else:
                commands.append(cmd)

        results: list[Result] = []
        for cmd in commands:
            results.append(self._maybe_dispatch(self._adapter.execute(cmd, self._connection)))
        return results

    def _drain(self, key: str, rule: Rule | None) -> list[Result]:
        if rule is None:
            return []
        current = self._state.get_overlap(key)
        drained = drain_transition(current, rule.overlap.policy)
        self._state.set_overlap(key, drained.new_state)
        if drained.decision != OverlapDecision.EXECUTE:
            return []
        pending = self._pending.pop(key, None)
        if pending is None:
            return []
        matched = self._match(pending)
        if matched is None:
            return []
        thread_id = pending.thread_id
        replay = StepResult(
            commands=[Typing(thread_id=thread_id), Host(handler_id=matched.handler_id)],
            matched_rule=matched,
            skipped_count=drained.skipped_count,
        )
        return self._execute(pending, replay) + self._drain(key, matched)

    def _maybe_dispatch(self, exec_result: Result) -> Result:
        if not exec_result.is_ok or self._transport is None:
            return exec_result
        sent = exec_result.value
        if isinstance(sent, Sent) and sent.raw.get("transport"):
            return self._transport.dispatch(sent)
        return exec_result

File: packages/typescript/src/interpreters/process.ts
/**
 * Process interpreter — self-host inbound.
 *
 * verify → parse → step → executeTurn → transport.
 * Webhook, poll, and cx.handle all call handleRaw.
 */
import * as Effect from "effect/Effect"
import type * as Layer from "effect/Layer"
import type { App } from "../core/app.ts"
import type { Command } from "../core/commands.ts"
import type { Connection } from "../core/connection.ts"
import { DecodeError, type CaspianError } from "../core/errors.ts"
import type { Event } from "../core/events.ts"
import {
  AdapterPort,
  type HostPort,
  type Sent,
  type StreamSink,
  type ThreadStore,
} from "../core/ports.ts"
import {
  makeMemoryInterpreter,
  type MemoryInterpreter,
} from "./memory.ts"
import { maybeDispatch, type Transport } from "./transport.ts"

export type HandleResult =
  | { readonly ok: true; readonly value: Sent }
  | { readonly ok: false; readonly error: CaspianError }

export type ProcessOptions = {
  readonly channelName?: string
  readonly secretToken?: string
  readonly secretHeader?: string
  readonly connection: Connection
  readonly adapter: Layer.Layer<AdapterPort>
  readonly host?: Layer.Layer<HostPort, never, ThreadStore>
  readonly transport?: Transport
}

export type ProcessInbound = {
  readonly secretHeader?: string
}

export type ProcessInterpreter = {
  readonly handle: (
    request: Request,
    inbound?: ProcessInbound,
  ) => Effect.Effect<Response>
  readonly handleRaw: (
    body: unknown,
    headers?: { readonly [key: string]: string },
  ) => Effect.Effect<ReadonlyArray<HandleResult>>
}

const tryJson = (body: string): unknown => {
  try {
    return JSON.parse(body)
  } catch {
    return body
  }
}

const jsonOf = (request: Request): Effect.Effect<unknown> =>
  Effect.tryPromise({
    try: () => request.json() as Promise<unknown>,
    catch: () => new Error("invalid json"),
  }).pipe(Effect.orElseSucceed((): unknown => ({})))

const headersOf = (request: Request): { readonly [key: string]: string } => {
  const headers: { [key: string]: string } = {}
  request.headers.forEach((value, key) => {
    headers[key] = value
  })
  return headers
}

const asHandleResult = (
  effect: Effect.Effect<Sent, import("../core/errors.ts").AdapterError>,
): Effect.Effect<HandleResult> =>
  effect.pipe(
    Effect.map((value): HandleResult => ({ ok: true, value })),
    Effect.catchAll((error): Effect.Effect<HandleResult> =>
      Effect.succeed({ ok: false, error }),
    ),
  )

export const makeProcessInterpreter = (
  app: App,
  options: ProcessOptions,
): Effect.Effect<ProcessInterpreter> =>
  Effect.gen(function* () {
    const streamSink: StreamSink | undefined =
      options.transport === undefined
        ? undefined
        : {
            can_stream: true,
            emit: (command) =>
              Effect.gen(function* () {
                const adapter = yield* AdapterPort
                const sent = yield* adapter.execute(command, options.connection)
                const dispatched = yield* maybeDispatch(options.transport, sent)
                return dispatched.message_id
              }).pipe(Effect.provide(options.adapter)),
          }

    // Commands are sent as they are produced, not collected and flushed after
    // the handler returns. The kernel emits the ack and the typing hint BEFORE
    // the Host command, and both exist to be seen while the agent thinks: with
    // a slow handler, batching shows the user nothing at all until it finishes.
    let liveResults: HandleResult[] | undefined

    const dispatchNow = (command: Command) =>
      Effect.gen(function* () {
        const adapter = yield* AdapterPort
        return yield* asHandleResult(
          adapter
            .execute(command, options.connection)
            .pipe(Effect.flatMap((sent) => maybeDispatch(options.transport, sent))),
        )
      }).pipe(Effect.provide(options.adapter))

    const memory: MemoryInterpreter = yield* makeMemoryInterpreter(app, {
      channelName: options.channelName ?? "",
      ...(options.host === undefined ? {} : { host: options.host }),
      ...(streamSink === undefined ? {} : { streamSink }),
      onProduce: (commands) =>
        Effect.gen(function* () {
          if (liveResults === undefined) return // not inside handleRaw
          for (const command of commands) {
            if (command.tag === "Host") continue // not a wire command
            liveResults.push(yield* dispatchNow(command))
          }
        }),
    })

    const runEvent = (event: Event) =>
      Effect.gen(function* () {
        const adapter = yield* AdapterPort
        yield* memory.run(event, adapter.overlapKey(event))
      }).pipe(Effect.provide(options.adapter))

    const handleRaw = (
      body: unknown,
      headers: { readonly [key: string]: string } = {},
    ): Effect.Effect<ReadonlyArray<HandleResult>> =>
      Effect.gen(function* () {
        const adapter = yield* AdapterPort
        const envelope = { body, headers }
        if (!adapter.verify(envelope, options.connection)) {
          return [
            {
              ok: false as const,
              error: new DecodeError({
                reason: "Webhook signature verification failed",
              }),
            },
          ]
        }
        // Webhooks deliver bytes; adapters expect the decoded object. Decode
        // here, AFTER verify: signatures are HMACs over the exact raw body, so
        // verify must see the original string. A non-JSON body (form-encoded
        // platforms) passes through unchanged for the adapter to handle.
        const decodedBody = typeof body === "string" ? tryJson(body) : body
        const parsed = yield* adapter.parse(decodedBody).pipe(Effect.either)
        if (parsed._tag === "Left") {
          return [{ ok: false as const, error: parsed.left }]
        }
        const results: HandleResult[] = []
        for (const event of parsed.right) {
          const ack = yield* asHandleResult(
            adapter.acknowledge(event, options.connection).pipe(
              Effect.flatMap((sent) => maybeDispatch(options.transport, sent)),
            ),
          )
          results.push(ack)
          liveResults = results
          yield* runEvent(event).pipe(Effect.catchAll(() => Effect.void))
          liveResults = undefined
        }
        return results
      }).pipe(Effect.provide(options.adapter))

    const handle = (
      request: Request,
      inbound?: ProcessInbound,
    ): Effect.Effect<Response> =>
      Effect.gen(function* () {
        if (options.secretToken !== undefined) {
          const secretHeader = inbound?.secretHeader ?? options.secretHeader
          if (secretHeader === undefined) {
            return new Response(null, { status: 401 })
          }
          const got = request.headers.get(secretHeader)
          if (got !== options.secretToken) {
            return new Response(null, { status: 401 })
          }
        }
        const raw = yield* jsonOf(request)
        const results = yield* handleRaw(raw, headersOf(request))
        const unauthorized = results.some(
          (result) =>
            !result.ok &&
            result.error._tag === "DecodeError" &&
            result.error.reason === "Webhook signature verification failed",
        )
        if (unauthorized) {
          return new Response(null, { status: 401 })
        }
        const twiml = results
          .filter((result) => result.ok)
          .map((result) =>
            result.ok && typeof result.value.raw.twiml === "string"
              ? result.value.raw.twiml
              : "",
          )
          .find((item) => item.length > 0)
        if (twiml !== undefined) {
          return new Response(twiml, {
            status: 200,
            headers: { "content-type": "text/xml" },
          })
        }
        return new Response(null, { status: 200 })
      })

    return { handle, handleRaw }
  })
